### DISCUSSION TITLE
Is dynamic library loading supported?

### DISCUSSION THREAD

**Discussion by IoIxD** [autc04/Retro68#262] [Q&A] [answered]
Curious because I know the compiler can compile and link to .a files statically but I've never tried dynamically, and I'm curious as to whether the compiler supports the actual Classic Mac OS dynamic library format?

**Comment by autc04** [comment, marked as answer]
For PowerPC, the linker supports shared libraries in XCOFF format. The actual Mac OS dynamic library format is PEF, and you can use the `MakePEF` and `MakeImport` tools to convert (`MakePEF` for XCOFF -> PEF, and `MakeImport` for PEF -> XCOFF stub library).

Following the tradition from AIX (which Apple based their PowerPC support on originally), both static and dynamic libraries are often stored in .a files. In fact, all the operating system functions are accessed via shared libraries (such as `InterfaceLib`, converted to `libInterfaceLib.a` by the `MakeImport` tool).


On 68K, Retro68 currently has no support for Shared Libraries.
Apple introduced support twice, once via the 'Apple Shared Library Manager', which no one ever used, and once via CFM68K which was introduced when 68K Macs were already obsolete.

**Comment by autc04** [comment]
Look in `Samples/SharedLibrary` for an example.

To load shared libraries at runtime (as opposed to at program startup as this example does), read Apple's documentation about the Code Fragment Manager to find out what the equivalents of UNIX `dlopen` and `dlsym` or Windows `LoadLibrary` and `GetProcAddress` are called.
